functorch/_src/vmap.py

```diff
@@ -366,4 +366,119 @@ def wrapped(*args, **kwargs):
             return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func)
         finally:
             _vmap_decrement_nesting()
+
     return wrapped
+
+
+def chunk_vmap(
+        func: Callable,
+        in_dims: in_dims_t = 0,
+        out_dims: out_dims_t = 0,
+        randomness: str = 'error',
+        chunks=2) -> Callable:
+    """
+    chunk_vmap is the vectorizing map using chunks of input data. Splitting data into chunks
+    can help to prevent out-of-memory issues. For more details about vectorizing map, see :func:`vjp`.
+
+    Args:
+        func (function): A Python function that takes one or more arguments.
+            Must return one or more Tensors.
+        in_dims (int or nested structure): Specifies which dimension of the
+            inputs should be mapped over. :attr:`in_dims` should have a
+            structure like the inputs. If the :attr:`in_dim` for a particular
+            input is None, then that indicates there is no map dimension.
+            Default: 0.
+        out_dims (int or Tuple[int]): Specifies where the mapped dimension
+            should appear in the outputs. If :attr:`out_dims` is a Tuple, then
+            it should have one element per output. Default: 0.
+        randomness (str): Specifies whether the randomness in this
+            vmap should be the same or different across batches. If 'different',
+            the randomness for each batch will be different. If 'same', the
+            randomness will be the same across batches. If 'error', any calls to
+            random functions will error. Default: 'error'. WARNING: this flag
+            only applies to random PyTorch operations and does not apply to
+            Python's random module or numpy randomness.
+        chunks (int): Number of chunks to use to split the input data. Default is 2.
+            If equals to 1 then :func:`vjp` is called.
+
+    Returns:
+        Returns a new "batched" function. It takes the same inputs as
+        :attr:`func`, except each input has an extra dimension at the index
+        specified by :attr:`in_dims`. It takes returns the same outputs as
+        :attr:`func`, except each output has an extra dimension at the index
+        specified by :attr:`out_dims`.
+    """
+    if randomness not in ['error', 'different', 'same']:
+        raise RuntimeError(f"Only allowed values for randomness are 'error', 'different', or 'same'. Got {randomness}")
```

Comment:
I just realized: randomness="same" doesn't actually work for chunked vmap, right? cc @samdow 

Let's assume we're doing the following:
```
def f(x):
  return torch.rand([])
  
x = torch.ones(4)

# returns the same value
vmap(f, randomness='same')(x)

# Returns torch.tensor([123, 123, 456, 456]). Read on for explanation.
chunk_vmap(f, randomness='same')(x)

# As implemented in this PR, chunk_vmap essentially calls the following:
chunk0 = vmap(f, randomness='same')(x[:2])
chunk1 = vmap(f, randomness='same')(x[2:])
result = torch.stack(chunk0, chunk1)
```

If we agree that is what this PR is doing, then one possible thing we can do right now:
- since we're tossing chunk_vmap in functorch.experimental.chunk_vmap, it's OK for this to not work and we can keep on thinking about what the semantics should be
- therefore it is OK to say we do not support randomness = same with chunk_vmap right now

But tbh I am not sure how we will support randomness = same with chunk_vmap in the future.

Reply:
For same we could maybe do something like:
```
        chunks_output = []
        rs = torch.get_rng_state()
        for flat_args in chunks_flat_args:
            batch_size = _validate_and_get_batch_size(flat_in_dims, flat_args)
            torch.set_rng_state(rs)
            chunks_output.append(
                _flat_vmap(batch_size, flat_in_dims, flat_args, args_spec, **kwargs)
            )

```

Reply:
Does this require the dimension being vmapped over to be evenly divisible by chunks?

Reply:
I thought about this some more. The approach here works!

> Do you think this can depend on dim values ?

In general, randomness depends on how many random numbers you're generating. So if we have the following:
```
def f(N):
  torch.manual_seed(0)
  _ = torch.randn(N)
  return torch.randn([])
```
Then `f(0)` and `f(1)` return different things. Each random number that is generated advance the global RNG state forward by one.

## The case I was concerned about

```py
def f(x):
  torch.randn_like(x)
  return torch.randn([])

x = torch.randn(3)
torch.manual_seed(0)
expected = vmap(f, randomness='same')(x)
print(expected)

x0, x1 = x.chunk(2)
torch.manual_seed(0)
y0 = vmap(f, randomness='same')(x0)
torch.manual_seed(0)
y1 = vmap(f, randomness='same')(x1)
result = torch.cat([y0, y1], 0)
print(result)
```
it turns out that when `randomness='same'`, then we end up generating the same amount of random numbers for the `randn_like` call (despite the bdims having different shapes), so the final torch.randn call always returns the same thing.

## In conclusion

Let's go with your approach. Please try to add additional tests for it with some examples like the one above)